### PR TITLE
 [`feat`] Implementing SmoothAP loss

### PR DESCRIPTION
**Pull request by ir2718**
Hi,

this PR implements the loss presented in the paper [Smooth-AP: Smoothing the Path Towards Large-Scale Image Retrieval](https://arxiv.org/abs/2007.12163). The default hyperparameters are chosen using the values that resemble a heavyside step function, as suggested by the authors.

Aside from the implementation, I also added a section in the docs and a test case which uses the original implementation to calculate the loss.

### CODE REVIEW

**Comment by ir2718** [comment]
Are there any plans on merging this or would you like me to make some changes?

File: src/pytorch_metric_learning/losses/smooth_ap.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        super().__init__(**kwargs)
+        c_f.assert_distance_type(self, CosineSimilarity)
+        self.temperature = temperature
+
+    def get_default_distance(self):
+        return CosineSimilarity()
+
+    # Implementation is based on the original repository:
+    # https://github.com/Andrew-Brown1/Smooth_AP/blob/master/src/Smooth_AP_loss.py#L87
+    def compute_loss(self, embeddings, labels, iices_tuple, ref_emb, ref_labels):
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/losses/smooth_ap.py:25]: typo, but also is there any possibility of supporting mining, or reference embeddings (ref_emb)?

File: src/pytorch_metric_learning/losses/smooth_ap.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                )
+            )
+
+        batch_size = embeddings.size(0)
+        num_classes_batch = batch_size // torch.unique(labels).size(0)
+
+        mask = 1.0 - torch.eye(batch_size)
+        mask = mask.unsqueeze(dim=0).repeat(batch_size, 1, 1)
+
+        sims = F.cosine_similarity(
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/losses/smooth_ap.py:52]: Would be nice to use self.distance, and loosen the restriction on self.distance being CosineSimilarity.

File: src/pytorch_metric_learning/losses/smooth_ap.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                sims_pos_ranks[i] / sims_ranks[i * g : (i + 1) * g, i * g : (i + 1) * g]
+            )
+            ap = ap + (pos_divide / g) / batch_size
+
+        loss = 1 - ap
+        return {
+            "loss": {
+                "losses": loss,
+                "indices": None,
+                "reduction_type": "already_reduced",
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/losses/smooth_ap.py:94]: Any possibility of returning an un-reduced loss?

Comment (ir2718) [reply src/pytorch_metric_learning/losses/smooth_ap.py]: Added.

File: src/pytorch_metric_learning/losses/smooth_ap.py (hunk 1 above)
Comment (ir2718) [reply src/pytorch_metric_learning/losses/smooth_ap.py]: I'm pretty sure I cannot add reference embeddings as the loss uses in-batch similarities for calculation. I've supported mining similar to the FastAP loss.

**Comment by ir2718** [comment]
@KevinMusgrave 

I think the changes resolve your concerns. Let me know if I missed anything.

**Comment by KevinMusgrave** [comment]
Thank you!
